Moving topics between streams

dobleuber opened the pull request:
https://github.com/zulip/zulip/issues/6427


**Feature Behaviour**
![Test topic - Zulip Dev - Zulip](https://user-images.githubusercontent.com/3765508/76370159-f124e480-6303-11ea-874b-eaf8ae35f27c.gif)

timabbott commented:
@dobleuber can you clean up the commit history?  Check out the Zulip commit guidelines for more details: https://zulip.readthedocs.io/en/latest/contributing/version-control.html

timabbott commented on zerver/lib/actions.py:
```diff
@@ -5900,3 +5900,61 @@ def do_delete_realm_export(user_profile: UserProfile, export: RealmAuditLog) ->
+    user_profile: UserProfile,
+    stream: Stream,
+    topic_name: str
+) -> List[Message]:
```
This line-wrapping doesn't match zulip's style.  Please look at nearby code and follow our conventions.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -5900,3 +5900,61 @@ def do_delete_realm_export(user_profile: UserProfile, export: RealmAuditLog) ->
+        user_profile=user_profile
+    )
+
+    recipient = stream.recipient
```
This variable is unnecessary.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -5900,3 +5900,61 @@ def do_delete_realm_export(user_profile: UserProfile, export: RealmAuditLog) ->
+        topic_name=topic_name,
+    )
+
+    return messages
```
See the above, but also I think this whole function could be like 2 lines of reasonable code; you possibly should just have the caller call `filter_by_topic_name_via_message`.

dobleuber replied:
I was getting this lint error:
```
avoid subject as a var at zerver/lib/actions.py line 5910:
message__subject__iexact=topic_name
```
So I called the method `filter_by_topic_name_via_message`

timabbott replied:
Yeah I just mean it could be:
```
query = UserMessage.objects.filter(user_profile=user_profile, message__recipient=stream.recipient)
return filter_by_topic_name_via_message(query=query, topic_name=topic_name)
```

Which might be better inlined unless we need it for tests.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -5900,3 +5900,61 @@ def do_delete_realm_export(user_profile: UserProfile, export: RealmAuditLog) ->
+def do_move_topic_messages(
+    user_profile: UserProfile,
+    stream_origin: Stream,
+    stream_destiny: Stream,
```
Call these `old_stream` and `new_stream`.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -5900,3 +5900,61 @@ def do_delete_realm_export(user_profile: UserProfile, export: RealmAuditLog) ->
+    user_messages = get_topic_messages(user_profile, stream_origin, topic_name)
+    recipient_destiny = get_stream_recipient(stream_destiny.id)
+
+    # We could use a bot for the breadcrump message too
```
Yeah, let's change this to a message sent by Notification Bot.

timabbott replied:
@mateuszmandera I just noticed we have these `get_stream_recipient` and `get_personal_recipient` helper functions in a few places that this code is copying; we should just remove those and use `stream.recipient` etc., right?  (And delete `get_recipient` and its cache as well, I think???)

In any case, for this PR, @dobleuber you can just use `new_stream.recipient`.

mateuszmandera replied:
@timabbott Yeah, we still have a bunch of places where we use those functions that haven't been changed. It should be pretty quick to do though and then we can remove the functions completely. I'll add it to my list.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -5900,3 +5900,61 @@ def do_delete_realm_export(user_profile: UserProfile, export: RealmAuditLog) ->
+
+    for user_message in user_messages:
+        user_message.message.recipient = recipient_destiny
+        user_message.message.save(update_fields=['recipient'])
```
This isn't going to work for caching and real-time sync reasons.  

See https://zulip.readthedocs.io/en/latest/subsystems/caching.html and https://zulip.readthedocs.io/en/latest/subsystems/events-system.html for details.

timabbott replied:
After some investigation, the caching situation might actually be fine, because `flush_message` ignores `update_fields` currently and just deletes any modified messages from the cache.

Real-time sync is going to be much more problematic; we'll need to extend the interface for "edit message" events to allow the stream to change.  It shouldn't be too bad because the problem is very similar to that of topic editing which we already do that correctly for.  I think ultimately what we'll want to do have this go through the edit message code path.

dobleuber replied:
Ok, I'll continue reviewing `do_update_message`. It does a lot of things, I'm not sure if I can reuse that method (adding a stream parameter), because we could have performance issues. I'm going to continue reviewing the code, if I have any questions I'll let you know, thanks

timabbott commented:
Thanks for working on this @dobleuber!  I pushed back to this PR a set of commits that rewrite this to use the `do_update_messages` infrastructure.  Some notes:
* The tests need some cleanup; we never hardcode IDs like 10 in them, instead using `user.id` substitutions.
* The use of `do_update_messages` partially fixes the message edit history for this feature.  We'll also need to add display code for it, which we can leave for follow-up work as long as the data is there.
* It also sets us up with just a bit of plumbing work in its `views.py` caller to provide this feature via the API rather than a management command, which both makes UI possible and doesn't require server shell access to use.  I think probably we should plan to delete the management command work and instead focus on making the .
* The `send_event` code for real-time sync needs work, both in this backend code (to decide what users to notify correctly) and on the frontend (to update the UI in `exports.update_messages` within `message_events.js`).  I'm OK with a potentially somewhat weird intermediate state, since one can reload browser windows, but we should at least make sure we've thought through how this might work.

The events thing may be hard for you to fix efficiently, but @dobleuber in your next pass, can you work on cleaning up the tests, changing the guard messages to use `narrow_url` to provide nice links?  Feel free to squash my new commit into yours; I think it's better squashed, but wanted to leave it separate just to make it easy for you to see what I did.  Also you could work on making `update_message_backend` support this code path as an realm admins-only feature.

dobleuber commented:
@timabbott I worked in your last comments.
Please let me know if you have any other comment, thanks for your help :-)

timabbott commented:
@dobleuber can you comment on which of the TODOs noted above you resolved, and how?

dobleuber commented:
* cleanup the tests: I refactor a little the tests to avoid using hardcoded ids and to do the tests independent of the environment.
* changing the guard messages to use narrow_url to provide nice links: I called the method `topic_narrow_url` to get the new topic url and added it to the message.
* I added a condition to allow this method be run just for realm admin users

Oh! I miss the method `update_message_backend`. Working on it. thanks

dobleuber commented:
* Updated the method `update_message_backend` to allow changing messages stream.
* Updated tests for api call to `update_message_backend`

dobleuber commented:
@timabbott Any feedback for this? meanwhile I am working in UI part:
![image](https://user-images.githubusercontent.com/3765508/75063260-87f34380-54b2-11ea-9afe-eb3be8f42f4b.png)

pirate commented:
I love that UI, nice work so far @dobleuber.

timabbott commented:
For UI, I think we should start with modifying the existing message editing UI where one can change topics to also have a "Stream" field; I think that will be a lot simpler as a way to start.

timabbott commented:
@dobleuber I just pushed an extra commit to this PR that does an incomplete version of a refactor that I think we should do before merging this; I didn't have time to clean up the tests (etc.), but basically, my hope is to eliminate both the management command and the extra function, and just have a single code path calling `do_update_message` with various arguments.

I also changed the code in a way that should fix some obviously broken behavior with the breadcrumb messages (various pieces were swapped).

You'll need to:
* Fix the linter/syntax/etc. errors; this is totally untested
* Rewrite the backend tests to use the API rather than calling `do_*` functions. (Or maybe just delete the `do_move_topic` tests, they may be basically duplicates of the API tests)
* Do some manual testing and confirm that the links we generate are actually correct.

timabbott commented on zerver/views/messages.py:
```diff
@@ -1582,11 +1583,63 @@ def update_message_backend(request: HttpRequest, user_profile: UserMessage,
+        sender = get_system_bot(settings.NOTIFICATION_BOT)
+
+        # Send a notification to the new stream that the topic was moved.
+        user_mention = "**_%s|%s**" % (user_profile.full_name, user_profile.id)
```
These links are missing `@s`, which is why the notifications look ugly like this: 
![image](https://user-images.githubusercontent.com/2746074/75937563-19ab6b00-5e3a-11ea-9f1c-ccbf9688a748.png)

timabbott commented on zerver/tests/test_messages.py:
```diff
@@ -3164,6 +3166,141 @@ def test_propagate_all_topics(self) -> None:
+        messages = get_topic_messages(user_profile, new_stream, "test")
+        self.assertEqual(len(messages), 0)
+
+        self.assert_json_error(result, "Streams must be public")
```
Generally we do `assert_json_error` immediately after the result is obtained for readability; ca you change that?

dobleuber replied:
Do you want to put the `assert_json_error` just after api call? Sure, I can change that.

timabbott commented on zerver/tests/test_messages.py:
```diff
@@ -3164,6 +3166,141 @@ def test_propagate_all_topics(self) -> None:
+            'propagate_mode': 'change_all',
+        })
+
+        messages = get_topic_messages(user_profile, stream, "test")
```
Add a comment above this noting that we expect the messages to remain in the original stream/topic.

timabbott commented on zerver/tests/test_messages.py:
```diff
@@ -3164,6 +3166,141 @@ def test_propagate_all_topics(self) -> None:
+        msg_id = self.send_stream_message(user_profile.email, stream.name,
+                                          topic_name="test", content="First")
+        self.send_stream_message(user_profile.email, stream.name,
+                                 topic_name="test", content="Second")
```
This first half of the function seems like it's common to all of these tests; perhaps we should move it to a common helper function that's called in common code?

dobleuber replied:
Sure I was trying to do that, but I am getting weird results. Do you have any example of this??

timabbott replied:
There's not much complicated to do, but note that you'll want to name your function not starting with `test_`, or it'll be run as a test itself.

timabbott commented on zerver/tests/test_messages.py:
```diff
@@ -3164,6 +3166,141 @@ def test_propagate_all_topics(self) -> None:
+        result = self.client_patch("/json/messages/" + str(msg_id), {
+            'message_id': msg_id,
+            'new_stream_id': new_stream.id,
+            'propagate_mode': 'change_all'
```
We should probably add a test for the other possible values of `propagate_mode`

timabbott commented on zerver/lib/actions.py:
```diff
@@ -4578,6 +4579,47 @@ def do_update_message(user_profile: UserProfile, message: Message, topic_name: O
+    if new_stream is not None:
+        # For simplicity, right now we only allow changing the stream,
+        # not also the topic at the same time.
+        assert topic_name is None
```
I'm now kinda tempted to just add backend support for changing `topic_name` at the same time; having seen this work, I think it's probably about the same amount of code (mostly, we need to change the outcome of the relevant test to verify the messages ended up at the new topic).  Can you take care of that?

dobleuber replied:
Sure, let me see.

timabbott commented on templates/zerver/help/move-a-topic.md:
```diff
@@ -0,0 +1,17 @@
+
+Just admin users can move topics between two public streams.
+
+### Move a topic
```
Let's call this "Move a topic to another stream".

timabbott commented on templates/zerver/help/move-a-topic.md:
```diff
@@ -0,0 +1,17 @@
+
+1. Select the destiny stream where the topic will be moved.
+
+{end_tabs}
```
I think probably we should make this a section of the existing "Rename a topic" article, rather than it's own article.  We might then decide to rename the new article as well to e.g. "Moving topics".

dobleuber replied:
Sure, let me see how to do that :-)

timabbott commented on static/js/unread.js:
```diff
@@ -413,6 +413,22 @@ exports.update_unread_topics = function (msg, event) {
     );
 };
 
+exports.update_unread_moved_topic = function (msg) {
```
Does this need to be different than `exports.update_unread_topics`?  It seems like we could just change `new_topic` to `msg.topic` in the previous function and not need to add a new function, unless I'm missing something.

dobleuber replied:
`new_topic` is undefined for moving topics, so I am not sure how to validate that on this case.

timabbott commented on static/js/unread.js:
```diff
@@ -413,6 +413,22 @@ exports.update_unread_topics = function (msg, event) {
+        msg.id
+    );
+};
+
```
Did you test the unread behavior after using `manage.py mark_all_message_unread` moves the unread counts correctly?

dobleuber replied:
No, let me try.

timabbott commented on static/js/stream_popover.js:
```diff
@@ -500,6 +533,27 @@ exports.register_topic_handlers = function () {
+    $('body').on('click', '.sidebar-popover-move-topic-messages', function (e) {
+        const topic_row = $(e.currentTarget);
+        const stream_id = topic_row.data('stream-id');
+        const topic_name = topic_row.data('topic-name');
```
We generally access via `.attr('data-stream-id')` as a codebase convention, because using jQuery's `.data()` to write doesn't update the attribute in the DOM, which can be super confusing.

timabbott commented on static/js/stream_popover.js:
```diff
@@ -500,6 +533,27 @@ exports.register_topic_handlers = function () {
+        const stream_row = $(e.currentTarget);
+        const stream_id = stream_row.data('stream-id');
+        const message_id = stream_row.data('message-id');
+        message_edit.show_topic_edit_spinner(stream_row);
```
Have you tested this spinner behavior looks right (e.g. by adding a `sleep(5)` in the backend endpoint)?

dobleuber replied:
It doesn't work. Let me double check it!

timabbott commented on static/js/stream_popover.js:
```diff
@@ -262,6 +275,26 @@ function build_starred_messages_popover(e) {
+        html: true,
+        trigger: "manual",
+        fixed: true,
+    });
```
The positioning seems like it can be a little weird :/

![image](https://user-images.githubusercontent.com/2746074/75938325-55473480-5e3c-11ea-8169-fa4411a80fdc.png)

dobleuber replied:
I see this on my local machine, what browser are you using? 
![image](https://user-images.githubusercontent.com/3765508/75941476-d236d600-5e5d-11ea-8f0e-b17d4703396a.png)

dobleuber replied:
Ohhh I got it!! let me fix it, thanks!

timabbott commented on static/js/stream_popover.js:
```diff
@@ -70,6 +76,13 @@ exports.hide_starred_messages_popover = function () {
     }
 };
 
+exports.hide_move_topic_to_stream_popover = function () {
```
Should this get called in `popovers.hide_all()`?    `exports.hide_all_except_sidebars` I think needs a line to call this, unless I'm missing something.

dobleuber replied:
I am not sure, the popover kept open whenever it lost the focus, I saw the others popovers, and I found that, how could I hide the popover when it lost the focus??

timabbott replied:
I would just call this from `exports.hide_all_except_sidebars`, which is the general function we call for hiding popovers in a range of situations where it makes sense to do so.

timabbott commented on static/templates/move_topic_to_stream.hbs:
```diff
@@ -0,0 +1,7 @@
+<ul class="nav nav-list topic-move_popover">
+    {{#each streams}}
+    <li>
+        <a href="#" class="popover-move-topic-to-stream" data-stream-id="{{ this.stream_id }}" data-message-id="{{ @root.message_id }}">{{this.name}}</a>
```
The `data-message-id` part could be put on the `ul` element rather than duplicating it on each stream.

timabbott replied:
Also, let's prefix the stream names with a `#`.

timabbott commented on static/js/message_edit.js:
```diff
@@ -766,4 +766,22 @@ exports.handle_narrow_deactivated = function () {
+        data: request,
+        success: function () {
+
+        },
```
The success handler should probably disable the spinner?  Probably doesn't need to do anything else.

dobleuber replied:
The spinner is not working... I'll need to fix it

timabbott commented on static/js/message_edit.js:
```diff
@@ -766,4 +766,22 @@ exports.handle_narrow_deactivated = function () {
+        },
+        error: function (xhr) {
+            ui_report.error(i18n.t("Error deleting message"), xhr,
+                            $("#delete-message-error"));
```
This string looks copy-pasted.

timabbott commented on static/js/message_events.js:
```diff
@@ -136,16 +137,23 @@ exports.update_messages = function update_messages(events) {
 
         const new_topic = util.get_edit_event_topic(event);
+        // TODO: notify and move messages whenever message's topics are moved
+        const {new_stream_id} =  event;
```
Fix the whitespace here.  Also is the TODO still incomplete?

dobleuber replied:
Nope, the TODO is fixed. I forgot to remove it

timabbott commented on static/js/message_events.js:
```diff
@@ -177,6 +185,21 @@ exports.update_messages = function update_messages(events) {
+                        then_select_id: current_id,
+                    };
+                    narrow.activate(operators, opts);
+                    changed_narrow = true;
```
I think this block should be combined with the previous one -- we want to only renarrow once if we later add stream+topic edits in a single event.  And we want the `selection_changed_topic` conditional (which I guess should be rebranded as `seleted_message_moved`)

timabbott commented on static/js/message_events.js:
```diff
@@ -177,6 +185,21 @@ exports.update_messages = function update_messages(events) {
+                    });
+                    const operators = new_filter.operators();
+                    const opts = {
+                        trigger: 'stream change',
```
This can be changed to 'message edit` when you combine with the previous block -- this value isn't parsed, it's just for debugging.

timabbott commented on static/js/message_events.js:
```diff
@@ -192,12 +215,22 @@ exports.update_messages = function update_messages(events) {
+                    // Update the unread counts; again, this must be called
+                    // before we call set_message_topic.
+                    unread.update_unread_moved_topic(msg);
+                }
```
These topic/stream blocks can probably be combined to just having the `msg.stream_id` line be not common code, after applying the other suggestions I've made.

timabbott commented on static/js/filter.js:
```diff
@@ -497,6 +497,17 @@ Filter.prototype = {
+            return new_term;
+        });
+        return new Filter(terms);
+    },
```
This can just replace `filter_with_new_topic`.  Can you extract this change as a prep commit that renames this and expands its functionality?  Then we can merge that ahead of the rest.

dobleuber replied:
Sure. :-)

dobleuber replied:
@timabbott how do I call this new PR??

dobleuber replied:
I created this PR: https://github.com/zulip/zulip/pull/14109

timabbott replied:
With `git rebase -i`, you can add a commit at the start of your PR for this sort of thing, but I suppose a new PR works too.

timabbott commented on zerver/tests/test_events.py:
```diff
@@ -746,7 +746,7 @@ def get_checker(check_gravatar: Validator) -> Validator:
         events = self.do_test(
-            lambda: do_update_message(self.user_profile, message, topic,
+            lambda: do_update_message(self.user_profile, message, None, topic,
                                       propagate_mode, content, rendered_content,
```
We will want to add a `test_events` test similar to the existing one for changing the stream.  

I opened https://github.com/zulip/zulip/issues/14101 for this; I think it is likely enough of a rabbit hole that we shouldn't go down it before merging this PR.

timabbott commented on zerver/views/messages.py:
```diff
@@ -1582,11 +1583,63 @@ def update_message_backend(request: HttpRequest, user_profile: UserMessage,
+        new_topic_link = "#**%s>%s**" % (new_stream.name, topic_name)
+        internal_send_stream_message(
+            old_stream.realm, sender, old_stream, old_topic_name,
+            "This topic was moved by %s to %s" % (user_mention, new_topic_link))
```
We should make sure that one of your tests verifies the `message.content` for these two messages.

Additionally, we should tag these strings for translation using `_()`.

timabbott commented:
@dobleuber huge thanks for your work on this!  I think it's going in a very promising direction.  I posted a long series of commits on how to avoid code duplication; once those are resolved, I think this will be at most a couple quick round trips away from being merged.

I still think we'll want to adjust the existing topic-editing UI to also support changing the stream, but I'm OK with leaving that until a bit later.

timabbott commented on static/templates/topic_sidebar_actions.hbs:
```diff
@@ -40,4 +40,13 @@
+    <li>
+        <a class="sidebar-popover-move-topic-messages" data-stream-id="{{ stream_id }}" data-topic-name="{{ topic_name }}">
+            <i class="fa fa-arrows" aria-hidden="true"></i>
+            {{#tr this}}Move all messages in <b>__topic_name__</b> to...{{/tr}}
```
Let's call this "Move <b>__topic_name__</b> to another stream."

timabbott commented on zerver/lib/actions.py:
```diff
@@ -4578,7 +4579,57 @@ def do_update_message(user_profile: UserProfile, message: Message, topic_name: O
+            # These fields have legacy field names.
+            event[ORIG_TOPIC] = old_topic_name
+            event[TOPIC_NAME] = topic_name
+            event[TOPIC_LINKS] = bugdown.topic_links(message.sender.realm_id, topic_name)
```
Would it be possible to avoid duplicating this code with the main `topic_name is not None` code path rather than doing `elif`? 

You're missing code this needs for the `edit_history_event`, and so I think the only complexity is making `changed_messages`  have `Set` semantics rather than `List` semantics.

Or am I missing something here?

dobleuber replied:
Ohhh!! I got it! sure! let me try again

dobleuber replied:
I updated that part again, Pls take a look :-)

onavhamid commented:
Wow - this would be a fantastic feature.  Thank you @dobleuber and @timabbott for pushing this along!

@timabbott , Would getting this merged automatically become available for https://zulipchat.com/ services as well?

timabbott commented:
I extracted the node tests changes as its own commit and merged it as 7702140521f7c8634f5f3a660264cdae4d44ee48.

timabbott commented:
OK, I just read this, and pushed an extra commit you'll need to fix up that adds  `notify_topic_moved` function as the way those notifications are sent (and also uses the `@_**Name**` silent mention syntax, which is important to avoid notifications for the person who moves the topic).  I also made a few edits in the main commit:
* Fixed the /help article's English and used standard macros in a few places.
* Changed the `zerver/lib/topic.py` code to not break other propagate modes.  There's a follow-up issue we should open for removing that "7 days" restriction noted in the TODO, but I very much don't want to bundle that sort of change with the rest of this.  @dobleuber can you open such an issue?

I think the following pieces are not done on the backend:
* The code move to extract `notify_topic_moved` needs to be fixed.  And maybe we should call it `notify_topic_moved_streams`.
* A few comments I'm about to post on the tests.

Once that's done, I'd prefer to split the commit so we can merge the backend/API piece, open any backend-specific follow-up issues, and be able to move forward with just the frontend left in this PR.

For the frontend, here are some initial comments:
* The manual `locale` change should be removed; i18n data is are synced from transifex.
* The click handler for `do_move_topic_button` should go in the main `register_click_handlers` mechanism, like our other click handlers, rather than being created every time we open the popover.
* That click handler is a little opaque and might be better written out more explicitly given that we only ever expect 2 parameters in that function.
* I'm fine with leaving adding a "stream" field to the message_edit form for a future issue.

timabbott commented on zerver/tests/test_messages.py:
```diff
@@ -3204,6 +3206,167 @@ def test_propagate_all_topics(self) -> None:
+        new_stream = prepared_topics[2]
+        msg_id = prepared_topics[3]
+
+        result = self.client_patch("/json/messages/" + str(msg_id), {
```
Please use `msg_id + 1` here; we want to verify that it only moved some of the messages.  Otherwise, this isn't a very good test of the `change_later` feature!

dobleuber replied:
Sorry, I don't catch it. What is the expected result for this test?? What do you mean by using `msg_id + 1`?

timabbott replied:
I mean move the 2nd message and later, not what you have moving the first message and later (which ends up being all the messages and thus doesn't test the precise behavior of change_later).

timabbott commented:
OK I also added 85c9ffd91cbc10d9d1432ad3d50e2884862910f4 to master, which allows us to do the relatively clean "squash me" commit I added at the top of the branch; I'm done working on this for the moment, so you can work without fear of conflicts.

timabbott commented:
@dobleuber just a quick ping on this -- we're really close to having this finished, and I'd love it if you can address that last batch of feedback so I can start merging it.

dobleuber commented:
Hi @timabbott. Sure, I hope to work on this soon, but right now I am busy. Maybe I could finish this week. Thanks for your support

dobleuber commented:
@timabbott I just created this issue related with your comment: https://github.com/zulip/zulip/issues/14492

timabbott commented:
@dobleuber OK I made the following changes:
* Changed the parameter for message editing to `stream_id`, rather than `new_stream_id`, for consistency with the existing `topic_name` parameter.
* Changed the backend tests to be cleaner and better match our style (including, critically, not assuming Iago's user ID will always be 10).
* Switched the mentions in the automatic notifications to be silent mentions.
* Rewrote much of the `message_events.js` code to 
* Added a Co-Authored-By for you since we seem to have squashed away your Git author info at some point along the way :).

The result is 843345dfee9c2e110e7add443f8dd3f9f6a8ace4 and 7990676583b5094e1a12572a782ece7ed57b401e.

The remaining commit is just the UI, which should be the simplest part.  I'm going to open a few follow-up issues before I get to looking at that, however.

timabbott commented:
I opened https://github.com/zulip/zulip/issues/14498 for the main known bug in this feature.

timabbott commented:
And I opened https://github.com/zulip/zulip/issues/14499 for tracking the future work we'll want to do once #14498 is done and we've gotten some experience playing with the feature and we have more user feedback on how they'd want to configure the permissions for this in their organization.

timabbott commented on static/js/stream_popover.js:
```diff
@@ -500,6 +512,35 @@ exports.register_topic_handlers = function () {
+        const stream_id = topic_row.attr('data-stream-id');
+        const topic_name = topic_row.attr('data-topic-name');
+        const message = home_msg_list.all_messages()
+            .find(m => m.topic === topic_name && m.stream_id === +stream_id);
```
This check doesn't seem reliable; we may not have any messages from the topic cached in the frontend.  I think the right fix is to fetch the latest message ID from `stream_topic_history.js`; that is guaranteed to have data given that this widget is visible at all.

@showell is there an existing function we can call, or does one need to be added to expose "any message ID from a given stream/topic appearing in the left sidebar"?

amanagr replied:
I don't think sending a message id is the right thing here. Shouldn't it be handled better on the server side using old and new streams and topics?

timabbott replied:
Well, for this UI, yes; but we'll also add this to "edit a single message" UI, and for that, the message ID is an important aspect of the meaning.  (Basically whenever `propagate_mode != all`)

amanagr replied:
Oh! That's good then.

timabbott commented on static/js/stream_popover.js:
```diff
@@ -500,6 +512,35 @@ exports.register_topic_handlers = function () {
+        if (old_topic_name && select_stream_id) {
+            message_edit.move_message_to_stream(message_id, select_stream_id, topic_name);
+            $('#move_topic_modal').modal('hide');
+        }
```
It looks like we're missing `stopPropagation/preventDefault` here.

timabbott replied:
Fixed this and pushed back here.

timabbott commented on static/js/message_edit.js:
```diff
@@ -854,4 +854,23 @@ exports.handle_narrow_deactivated = function () {
     }
 };
 
+exports.move_message_to_stream = function (message_id, stream_id, topic_name) {
```
Let's rename this to `exports.move_topic_containing_message_to_stream`, for clarity about what it does.

And maybe add a comment right here making clear the `stream_id` and `topic_name` are the NEW values, and the old ones are wherever the message is located, since that may not be intuitive.

timabbott commented on static/js/message_edit.js:
```diff
@@ -854,4 +854,23 @@ exports.handle_narrow_deactivated = function () {
+        },
+        error: function (xhr) {
+            ui_report.error(i18n.t("Error moving the topic"), xhr,
+                            $("#home-error"));
```
Can you post a screenshot of this appearing?

amanagr replied:
The cross icon in mission from the error popup.

timabbott replied:
I think it's probably good enough to just mark it to disappear after a few seconds.

timabbott commented on static/js/stream_popover.js:
```diff
@@ -262,6 +263,17 @@ function build_starred_messages_popover(e) {
 }
 
+function build_move_topic_to_stream_popover(e, message_id, stream_id, topic_name) {
+    const streams = stream_data.subscribed_subs().filter(s => s.stream_id !== stream_id);
```
Let's call this variable `available_streams`.  

I'll note that we currently include private streams in this list.

amanagr replied:
Done. Added a TODO for keyboard navigation of the dropdown.

amanagr replied:
Do we want to exclude private streams?

timabbott replied:
Well, I think there's a product question about whether it's correct to limit it to just public streams.  We should make a decision on that and adjust accordingly.  Initially we did it just to manage complexity.

timabbott commented on static/js/stream_popover.js:
```diff
@@ -500,6 +512,38 @@ exports.register_topic_handlers = function () {
+
+        if (old_topic_name.trim() === topic_name.trim()) {
+            topic_name = undefined;
+        }
```
Maybe add a comment here noting that this is intended to request no change in the topic if it didn't change?

amanagr replied:
Done

timabbott commented on templates/zerver/help/rename-a-topic.md:
```diff
@@ -26,3 +28,24 @@ for the details on when topic editing is allowed.
+
+1. Select **Move <topic name\> to another stream**.
+
+1. Select the destination stream (and if desired, enter a new topic name).
```
Let's just call it "a new topic" here.

amanagr replied:
Done.

timabbott commented:
OK, I posted a batch of mostly small comments; the main thing we need to fix here is the computation of a message ID, which hopefully @showell can help with.

timabbott commented on static/js/stream_popover.js:
```diff
@@ -262,6 +263,17 @@ function build_starred_messages_popover(e) {
+
+    $("#move-a-topic-modal-holder").html(render_move_topic_to_stream(args));
+    $('#move_topic_modal').modal('show');
+    e.stopPropagation();
```
We can delete this `stopPropagation`; it's in the caller (and should be there)

timabbott commented:
Closing in favor of #14844; @amanagr you still still read the last few comments here.  Thanks @dobleuber!

timabbott commented:
The UI for this was merged today.  Thanks again for all the work on this @dobleuber!
